### PR TITLE
Unify BatchReader and TraceReader traits.

### PR DESCRIPTION
**Pull request by ryzhyk**
This PR resolves issue #68.

### CODE REVIEW

File: src/operator/aggregate.rs [hunk 1]
```diff
@@ -178,19 +183,32 @@ where
...
+    Z: IndexedZSet<R = O::R> + 'static,
+    F: Fn(&Z::Key, &mut [(&Z::Val, Z::R)]) -> O::Key + 'static,
     O: Clone + ZSet + 'static,
     O::R: ZRingValue,
 {
-    fn eval(&mut self, i: &I) -> O {
+    fn eval(&mut self, i: &Z) -> O {
         let mut elements = Vec::with_capacity(i.len());
         let mut cursor = i.cursor();
+        let mut vals: Vec<(&Z::Val, Z::R)> = Vec::new();
```
Comment (Kixiron) [inline src/operator/aggregate.rs:196]: Should we pre-allocate this with the number of keys or values?

File: src/operator/aggregate.rs [hunk 2]
```diff
@@ -314,18 +341,13 @@ mod test {
...
                         let mut result: isize = 0;
 
-                        while cursor.val_valid(storage) {
-                            let v = cursor.val(storage);
-                            let w = cursor.weight(storage);
-                            result += (*v as isize) * w;
-                            cursor.step_val(storage);
+                        for (v, w) in vals.into_iter() {
+                            result += (**v as isize) * (*w);
                         }
```
Comment (Kixiron) [inline src/operator/aggregate.rs:349]: Github won't let me make a suggestion, but here it is
```rust
let result = vals.iter().map(|&(&v, w)| v * w).sum();
```

File: src/operator/aggregate.rs [hunk 3]
```diff
@@ -382,20 +404,16 @@ mod test {
...
-                     -> (usize, usize) {
+                    let min = |key: &usize, vals: &mut [(&usize, isize)]| -> (usize, usize) {
                         let mut result = usize::MAX;
 
-                        while cursor.val_valid(storage) {
-                            let v = cursor.key(storage);
-                            if v < &result {
+                        for (v, _) in vals.into_iter() {
+                            if *v < &result {
                                 result = v.clone();
```
Comment (Kixiron) [inline src/operator/aggregate.rs:412]:
```suggestion
                        for &(&v, _) in vals.iter() {
                            if v < result {
                                result = v;
```

File: src/trace/ord/val_batch.rs [hunk 4]
```diff
@@ -393,12 +393,17 @@ where
             self.cursor.child.child.step(&storage.layer.vals.vals);
         }
     }
-    fn weight<'a>(&self, storage: &'a Self::Storage) -> &'a R
+    fn weight(&mut self, storage: &Self::Storage) -> R
```
Comment (Kixiron) [inline src/trace/ord/val_batch.rs:396]: Looks like we clone within pretty much every `.weight()` call, maybe it should return a reference and allow the caller to decide whether or not it should be cloned

File: src/operator/aggregate.rs [hunk 5]
```diff
@@ -178,19 +183,32 @@ where
...
+                let w = cursor.weight(i);
+                // Skip values with weight zero.
+                if !w.is_zero() {
+                    vals.push((cursor.val(i), w));
+                }
+                cursor.step_val(i);
+            }
+            // Skip keys that only contain values with weight zero.
+            if !vals.is_empty() {
+                elements.push((((self.agg_func)(cursor.key(i), &mut vals), ()), Z::R::one()));
```
[the remark this answers is not in the document]
Comment (ryzhyk) [reply src/operator/aggregate.rs:209]: That might actually work, thanks for the idea.

File: src/operator/aggregate.rs (hunk 1 above)
Comment (ryzhyk) [reply src/operator/aggregate.rs:196]: At this point we have no insight about the average/max size of a group. But I guess there's not much harm in preallocating `delta.len()` elements.

File: src/trace/ord/val_batch.rs (hunk 4 above)
Comment (ryzhyk) [reply src/trace/ord/val_batch.rs:396]: Problem is, the result sometimes needs to be computed, but I can probably make it work...

Comment (ryzhyk) [reply src/trace/ord/val_batch.rs:396]: Actually, no, not in a nice way.

**Pull request merged**